// react-native-web/dist/cjs/modules/createDOMProps/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _AccessibilityUtil = _interopRequireDefault(require("../AccessibilityUtil"));

var _css = _interopRequireDefault(require("../../exports/StyleSheet/css"));

var _StyleSheet = _interopRequireDefault(require("../../exports/StyleSheet"));

var _styleResolver2 = _interopRequireDefault(require("../../exports/StyleSheet/styleResolver"));

var _constants = require("../../exports/StyleSheet/constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var emptyObject = {};
var hasOwnProperty = Object.prototype.hasOwnProperty; // Reset styles for heading, link, and list DOM elements

var classes = _css.default.create({
  reset: {
    backgroundColor: 'transparent',
    color: 'inherit',
    font: 'inherit',
    listStyle: 'none',
    margin: 0,
    textAlign: 'inherit',
    textDecoration: 'none'
  },
  cursor: {
    cursor: 'pointer'
  }
}, _constants.STYLE_GROUPS.classicReset);

var pointerEventsStyles = _StyleSheet.default.create({
  auto: {
    pointerEvents: 'auto'
  },
  'box-none': {
    pointerEvents: 'box-none'
  },
  'box-only': {
    pointerEvents: 'box-only'
  },
  none: {
    pointerEvents: 'none'
  }
});

var defaultStyleResolver = function defaultStyleResolver(style, classList) {
  return _styleResolver2.default.resolve(style, classList);
};

var createDOMProps = function createDOMProps(component, props, styleResolver) {
  if (!styleResolver) {
    styleResolver = defaultStyleResolver;
  }

  if (!props) {
    props = emptyObject;
  }

  var _props = props,
      accessibilityLabel = _props.accessibilityLabel,
      accessibilityLiveRegion = _props.accessibilityLiveRegion,
      accessibilityRelationship = _props.accessibilityRelationship,
      accessibilityState = _props.accessibilityState,
      accessibilityValue = _props.accessibilityValue,
      accessible = _props.accessible,
      classList = _props.classList,
      providedDisabled = _props.disabled,
      importantForAccessibility = _props.importantForAccessibility,
      nativeID = _props.nativeID,
      pointerEvents = _props.pointerEvents,
      providedStyle = _props.style,
      testID = _props.testID,
      accessibilityRole = _props.accessibilityRole,
      unstable_ariaSet = _props.unstable_ariaSet,
      unstable_dataSet = _props.unstable_dataSet,
      domProps = _objectWithoutPropertiesLoose(_props, ["accessibilityLabel", "accessibilityLiveRegion", "accessibilityRelationship", "accessibilityState", "accessibilityValue", "accessible", "classList", "disabled", "importantForAccessibility", "nativeID", "pointerEvents", "style", "testID", "accessibilityRole", "unstable_ariaSet", "unstable_dataSet"]);

  var disabled = accessibilityState != null && accessibilityState.disabled === true || providedDisabled;

  var role = _AccessibilityUtil.default.propsToAriaRole(props); // unstable_ariaSet


  if (unstable_ariaSet != null) {
    for (var prop in unstable_ariaSet) {
      if (hasOwnProperty.call(unstable_ariaSet, prop)) {
        var value = unstable_ariaSet[prop];

        if (value != null) {
          domProps["aria-" + prop] = value;
        }
      }
    }
  } // unstable_dataSet


  if (unstable_dataSet != null) {
    for (var _prop in unstable_dataSet) {
      if (hasOwnProperty.call(unstable_dataSet, _prop)) {
        var _value = unstable_dataSet[_prop];

        if (_value != null) {
          domProps["data-" + _prop] = _value;
        }
      }
    }
  } // accessibilityLabel


  if (accessibilityLabel != null) {
    domProps['aria-label'] = accessibilityLabel;
  } // accessibilityLiveRegion


  if (accessibilityLiveRegion != null) {
    domProps['aria-live'] = accessibilityLiveRegion === 'none' ? 'off' : accessibilityLiveRegion;
  } // accessibilityRelationship


  if (accessibilityRelationship != null) {
    for (var _prop2 in accessibilityRelationship) {
      var _value2 = accessibilityRelationship[_prop2];

      if (_value2 != null) {
        domProps["aria-" + _prop2] = _value2;
      }
    }
  } // accessibilityRole


  if (role != null) {
    domProps.role = role;
  } // accessibilityState


  if (accessibilityState != null) {
    for (var _prop3 in accessibilityState) {
      var _value3 = accessibilityState[_prop3];

      if (_value3 != null) {
        if (_prop3 === 'disabled' || _prop3 === 'hidden') {
          if (_value3 === true) {
            domProps["aria-" + _prop3] = _value3; // also set prop directly to pick up host component behaviour

            domProps[_prop3] = _value3;
          }
        } else {
          domProps["aria-" + _prop3] = _value3;
        }
      }
    }
  } // accessibilityValue


  if (accessibilityValue != null) {
    for (var _prop4 in accessibilityValue) {
      var _value4 = accessibilityValue[_prop4];

      if (_value4 != null) {
        domProps["aria-value" + _prop4] = _value4;
      }
    }
  } // legacy fallbacks


  if (importantForAccessibility === 'no-hide-descendants') {
    domProps['aria-hidden'] = true;
  }

  if (disabled === true) {
    domProps['aria-disabled'] = true;
    domProps.disabled = true;
  } // FOCUS
  // Assume that 'link' is focusable by default (uses <a>).
  // Assume that 'button' is not (uses <div role='button'>) but must be treated as such.


  var isInteractiveElement = role === 'link' || component === 'a' || component === 'button' || component === 'input' || component === 'select' || component === 'textarea';
  var focusable = !disabled && importantForAccessibility !== 'no' && importantForAccessibility !== 'no-hide-descendants';

  if (isInteractiveElement) {
    if (accessible === false || !focusable) {
      domProps.tabIndex = '-1';
    } else {
      domProps['data-focusable'] = true;
    }
  } else if (_AccessibilityUtil.default.buttonLikeRoles[role] || role === 'textbox') {
    if (accessible !== false && focusable) {
      domProps['data-focusable'] = true;
      domProps.tabIndex = '0';
    }
  } else {
    if (accessible === true && focusable) {
      domProps['data-focusable'] = true;
      domProps.tabIndex = '0';
    }
  } // STYLE


  var reactNativeStyle = _StyleSheet.default.compose(pointerEvents && pointerEventsStyles[pointerEvents], providedStyle); // Additional style resets for interactive elements


  var needsCursor = (role === 'button' || role === 'link') && !disabled;
  var needsReset = component === 'a' || component === 'button' || component === 'li' || component === 'ul' || role === 'heading'; // Classic CSS styles

  var finalClassList = [needsReset && classes.reset, needsCursor && classes.cursor, classList]; // Resolve styles

  var _styleResolver = styleResolver(reactNativeStyle, finalClassList),
      className = _styleResolver.className,
      style = _styleResolver.style;

  if (className != null && className !== '') {
    domProps.className = className;
  }

  if (style) {
    domProps.style = style;
  } // OTHER
  // Native element ID


  if (nativeID != null) {
    domProps.id = nativeID;
  } // Link security
  // https://mathiasbynens.github.io/rel-noopener/
  // Note: using "noreferrer" doesn't impact referrer tracking for https
  // transfers (i.e., from https to https).


  if (component === 'a' && domProps.target === '_blank') {
    domProps.rel = (domProps.rel || '') + " noopener noreferrer";
  } // Automated test IDs


  if (testID != null) {
    domProps['data-testid'] = testID;
  } // Keyboard accessibility
  // Button-like roles should trigger 'onClick' if SPACE or ENTER keys are pressed.
  // Button-like roles should not trigger 'onClick' if they are disabled.


  if (domProps['data-focusable']) {
    var onClick = domProps.onClick;

    if (onClick != null) {
      if (disabled) {
        domProps.onClick = undefined;
      } else if (!isInteractiveElement) {
        var onKeyDown = domProps.onKeyDown;

        domProps.onKeyDown = function (e) {
          if (!e.isDefaultPrevented() && (e.key === 'Enter' || e.key === ' ')) {
            e.preventDefault();

            if (onKeyDown != null) {
              onKeyDown(e);
            }

            onClick(e);
          }
        };
      }
    }
  }

  return domProps;
};

var _default = createDOMProps;
exports.default = _default;
module.exports = exports.default;
